### PR TITLE
Replace CreateStreamOnHGlobal with a custom COM Memory Stream

### PR DESCRIPTION
**Pull request by pharring**
CreateStreamOnHGlobal turns out to have terrible performance for the access pattern used by DiaSymReader (lots of small "SetSize" calls)

So, now we implement the stream in managed code. The main advantage of this implementation is that it uses byte[] chunks rather than trying to keep a contiguous memory block.

Also, we redefine the IStream interface with a small tweak to "Write" to avoid allocating byte[] when called from native code.

### CODE REVIEW

**Comment by pharring** [comment]
@tmat @agocke @jaredpar for review.

File: src/Compilers/Core/Portable/PEWriter/ComMemoryStream.cs [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                    {
+                        stream.WriteByte(0);
+                    }
+                }
+
+                cb -= bytesToCopy;
+            }
+        }
+
+        unsafe void IUnsafeComStream.Read(byte[] pv, int cb, IntPtr pcbRead)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/PEWriter/ComMemoryStream.cs]: private

Comment (pharring) [inline src/Compilers/Core/Portable/PEWriter/ComMemoryStream.cs]: You can't have accessibility modifiers on an explicit interface implementation.

**Comment by pharring** [comment]
The performance improvement is quite impressive (or rather, it shows how slow CreateStreamOnHGlobal was):
C# compiler throughput (Roslyn snapshot):
Before: 41.559 seconds CPU time, 25.010 seconds elapsed
After:30.882 seconds CPU time, 14.519 seconds elapsed

**Comment by pharring** [comment]
@tmat @jaredpar ping? Looking for sign-off so I can get this into a perf run today.

**Comment by tmat** [comment]
:+1:

**Comment by agocke** [comment]
Won't this require the same PInvoke overhead that we saw the first time we stopped using a native stream?

**Comment by agocke** [comment]
LGTM

**Comment by pharring** [comment]
@agocke That's been (somewhat) mitigated by using an IntPtr for Write. Once we get new (official) perf traces, I'll look to see if I can squeeze any more out of it.

**Pull request merged**